File: rust/src/orchestration.rs
//! callwire orchestration — zero-config multi-language process management.
//!
//! # Two modes, selected automatically
//!
//! **Orchestrator mode** (default, `CALLWIRE_SPAWNED` is unset):
//! - Reads `callwire.toml` from the current directory.
//! - Starts a dynamic registry on a random OS-assigned port.
//! - Spawns declared worker services as child processes with:
//!   - `CALLWIRE_SPAWNED=1`
//!   - `CALLWIRE_REGISTRY=127.0.0.1:<port>`
//! - Returns an [`OrchestratorGuard`]; dropping it kills all children.
//!
//! **Worker mode** (`CALLWIRE_SPAWNED=1`):
//! - Skips spawning.
//! - Starts the local RPC server on a random OS port.
//! - Connects to `CALLWIRE_REGISTRY`, registers all locally-exported functions.
//! - Spawns an orphan-detection task that calls `std::process::exit(0)` if
//!   the parent process dies unexpectedly.

use std::collections::HashMap;
use std::io::Write;
use std::path::Path;
use std::process::{Child, Command};
use std::time::Duration;
use tokio::net::TcpListener;

// ── TOML parsing ─────────────────────────────────────────────────────────────

#[derive(Debug, serde::Deserialize)]
struct CallwireToml {
    services: Option<HashMap<String, ServiceConfig>>,
}

#[derive(Debug, serde::Deserialize)]
struct ServiceConfig {
    dev_cmd: Option<String>,
    prod_cmd: Option<String>,
}

fn load_toml() -> anyhow::Result<CallwireToml> {
    let path = Path::new("callwire.toml");
    if !path.exists() {
        return Ok(CallwireToml { services: None });
    }
    let content = std::fs::read_to_string(path)?;
    let config: CallwireToml = toml::from_str(&content)?;
    Ok(config)
}

// ── PID file management ──────────────────────────────────────────────────────

fn kill_stale_pids() {
    let pid_file = Path::new(".callwire/pids");
    if !pid_file.exists() {
        return;
    }
    if let Ok(content) = std::fs::read_to_string(pid_file) {
        for line in content.lines() {
            if let Ok(pid) = line.trim().parse::<u32>() {
                // Best-effort: send SIGTERM
                #[cfg(unix)]
                unsafe {
                    libc::kill(pid as libc::pid_t, libc::SIGTERM);
                }
            }
        }
    }
    let _ = std::fs::remove_file(pid_file);
}

fn write_pid_file(children: &[Child]) {
    let _ = std::fs::create_dir_all(".callwire");
    if let Ok(mut f) = std::fs::File::create(".callwire/pids") {
        for child in children {
            let _ = writeln!(f, "{}", child.id());
        }
    }
}

// ── free port helper ─────────────────────────────────────────────────────────

async fn free_port() -> std::io::Result<u16> {
    let listener = TcpListener::bind("127.0.0.1:0").await?;
    Ok(listener.local_addr()?.port())
}

// ── OrchestratorGuard ────────────────────────────────────────────────────────

/// Returned by [`init`] in orchestrator mode.  
/// Dropping this value terminates all spawned worker processes.
pub struct OrchestratorGuard {
    children: Vec<Child>,
}

impl OrchestratorGuard {
    /// Explicitly shut down all workers. Equivalent to dropping the guard.
    pub fn shutdown(mut self) {
        self.terminate_children();
    }

    fn terminate_children(&mut self) {
        for child in &mut self.children {
            let _ = child.kill();
        }
        for child in &mut self.children {
            let _ = child.wait();
        }
        let _ = std::fs::remove_file(".callwire/pids");
    }
}

impl Drop for OrchestratorGuard {
    fn drop(&mut self) {
        self.terminate_children();
    }
}

// ── public API ────────────────────────────────────────────────────────────────

/// Initialise Callwire orchestration.
///
/// Call once at application startup (e.g. in `#[tokio::main]`, an Axum
/// lifespan, or an Actix `on_start`).
///
/// Returns `Some(OrchestratorGuard)` in orchestrator mode (drop to shut down
/// workers), or `None` in worker mode.
///
/// # Example
/// ```no_run
/// #[tokio::main]
/// async fn main() {
///     let _guard = callwire::init().await.expect("callwire init failed");
///     // your server logic
/// }
/// ```
pub async fn init() -> anyhow::Result<Option<OrchestratorGuard>> {
    if std::env::var("CALLWIRE_SPAWNED").as_deref() == Ok("1") {
        init_as_worker().await?;
        return Ok(None);
    }
    let guard = init_as_orchestrator().await?;
    Ok(Some(guard))
}

// ── worker mode ──────────────────────────────────────────────────────────────

async fn init_as_worker() -> anyhow::Result<()> {
    let registry_addr = std::env::var("CALLWIRE_REGISTRY")
        .map_err(|_| anyhow::anyhow!("CALLWIRE_REGISTRY not set in worker mode"))?;

    // Bind to a random OS port
    let listener = TcpListener::bind("127.0.0.1:0").await?;
    let worker_addr = listener.local_addr()?.to_string();
    eprintln!("[callwire] Worker serving on {}", worker_addr);

    // Start the RPC accept loop in the background
    let (_tx, rx) = tokio::sync::watch::channel(false);
    tokio::spawn(crate::server::run_accept_loop(listener, rx));

    // Brief grace period
    tokio::time::sleep(Duration::from_millis(150)).await;

    // Collect exported function names from the global REGISTRY
    let func_names: Vec<String> = {
        let reg = crate::server::REGISTRY.lock().unwrap();
        reg.keys()
            .filter(|k| !k.starts_with("callwire."))
            .cloned()
            .collect()
    };

    // Register each with the parent registry
    let client = crate::client::Client::connect(&registry_addr).await?;
    for name in &func_names {
        if let Err(e) = client
            .import::<(), _>("callwire.register", &(name.as_str(), worker_addr.as_str()))
            .await
        {
            eprintln!("[callwire] Warning: could not register '{}': {}", name, e);
        }
    }
    eprintln!("[callwire] Worker registered {:?} → {}", func_names, worker_addr);

    // Orphan detection: exit if parent dies
    let parent_pid = std::os::unix::process::parent_id();
    tokio::spawn(async move {
        loop {
            tokio::time::sleep(Duration::from_secs(2)).await;
            #[cfg(unix)]
            {
                let current = std::os::unix::process::parent_id();
                if current == 1 && current != parent_pid {
                    eprintln!("[callwire] Parent process gone — worker exiting");
                    std::process::exit(0);
                }
            }
        }
    });

    Ok(())
}

// ── orchestrator mode ─────────────────────────────────────────────────────────

async fn init_as_orchestrator() -> anyhow::Result<OrchestratorGuard> {
    let config = load_toml()?;
    let services = config.services.unwrap_or_default();

    if services.is_empty() {
        return Ok(OrchestratorGuard { children: vec![] });
    }

    // Kill PIDs from a previous run
    kill_stale_pids();

    // Start the dynamic registry on a random port
    let reg_port = free_port().await?;
    let registry_addr = format!("127.0.0.1:{}", reg_port);

    // Register callwire.register and callwire.discover on this process
    crate::registry::start_embedded_registry().await;
    crate::server::serve_on(&registry_addr).await?;
    eprintln!("[callwire] Registry listening on {}", registry_addr);
    unsafe {
        std::env::set_var("CALLWIRE_REGISTRY", &registry_addr);
    }

    let is_prod = std::env::var("CALLWIRE_ENV")
        .map(|v| v.to_lowercase() == "prod")
        .unwrap_or(false);

    let mut children = Vec::new();

    for (name, svc) in &services {
        let cmd = if is_prod {
            svc.prod_cmd.as_deref().or(svc.dev_cmd.as_deref())
        } else {
            svc.dev_cmd.as_deref().or(svc.prod_cmd.as_deref())
        };

        let Some(cmd) = cmd else {
            eprintln!("[callwire] Warning: service '{}' has no command — skipping", name);
            continue;
        };

        let child = Command::new("sh")
            .arg("-c")
            .arg(cmd)
            .env("CALLWIRE_SPAWNED", "1")
            .env("CALLWIRE_REGISTRY", &registry_addr)
            .spawn();

        match child {
            Ok(c) => {
                eprintln!("[callwire] Spawned '{}' (PID {}): {}", name, c.id(), cmd);
                children.push(c);
            }
            Err(e) => {
                eprintln!("[callwire] Failed to spawn '{}': {}", name, e);
            }
        }
    }

    write_pid_file(&children);

    // Wait for workers to come up
    let wait_ms = std::cmp::min(1500 * children.len().max(1) as u64, 5000);
    tokio::time::sleep(Duration::from_millis(wait_ms)).await;

    eprintln!("[callwire] Orchestrator ready — registry at {}", registry_addr);

    Ok(OrchestratorGuard { children })
}

File: ts/src/orchestration.ts
import * as fs from 'fs';
import * as net from 'net';
import * as path from 'path';
import * as child_process from 'child_process';
import { Server } from './server';
import { Client } from './client';

// ── TOML mini-parser (no external dep for the subset we need) ─────────────────

interface ServiceConfig {
  dev_cmd?: string;
  prod_cmd?: string;
}

interface CallwireToml {
  services: Record<string, ServiceConfig>;
}

function parseToml(): CallwireToml {
  const tomlPath = path.resolve('callwire.toml');
  if (!fs.existsSync(tomlPath)) return { services: {} };

  const content = fs.readFileSync(tomlPath, 'utf-8');
  const result: CallwireToml = { services: {} };
  let currentService: string | null = null;

  for (const rawLine of content.split('\n')) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;

    // Section header [services.<name>]
    const sectionMatch = line.match(/^\[services\.(.+)\]$/);
    if (sectionMatch) {
      currentService = sectionMatch[1];
      result.services[currentService] = {};
      continue;
    }
    // Skip other headers
    if (line.startsWith('[')) { currentService = null; continue; }

    if (!currentService) continue;
    const eqIdx = line.indexOf('=');
    if (eqIdx === -1) continue;
    const key = line.slice(0, eqIdx).trim();
    let val = line.slice(eqIdx + 1).trim();
    if (val.startsWith('"') && val.endsWith('"')) val = val.slice(1, -1);

    if (key === 'dev_cmd') result.services[currentService].dev_cmd = val;
    if (key === 'prod_cmd') result.services[currentService].prod_cmd = val;
  }

  return result;
}

// ── helpers ────────────────────────────────────────────────────────────────

function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = net.createServer();
    srv.listen(0, '127.0.0.1', () => {
      const addr = srv.address() as net.AddressInfo;
      const port = addr.port;
      srv.close(() => resolve(port));
    });
    srv.on('error', reject);
  });
}

function killStalePids() {
  const pidFile = path.resolve('.callwire', 'pids');
  if (!fs.existsSync(pidFile)) return;
  try {
    const pids = fs.readFileSync(pidFile, 'utf-8').split('\n').filter(Boolean);
    for (const pid of pids) {
      // Negative PID targets the process group (see the detached: true
      // comment at spawn time) so a stale hot-reload PID takes its whole
      // worker tree down, not just the shell wrapper.
      try { process.kill(-Number(pid), 'SIGTERM'); } catch {
        try { process.kill(Number(pid), 'SIGTERM'); } catch { /* already gone */ }
      }
    }
    fs.unlinkSync(pidFile);
  } catch { /* ignore */ }
}

function writePidFile(procs: child_process.ChildProcess[]) {
  fs.mkdirSync('.callwire', { recursive: true });
  const pids = procs.map(p => String(p.pid)).join('\n');
  fs.writeFileSync(path.join('.callwire', 'pids'), pids + '\n');
}

// ── OrchestratorHandle ────────────────────────────────────────────────────────

export class OrchestratorHandle {
  private procs: child_process.ChildProcess[];
  private registryServer: Server;

  constructor(procs: child_process.ChildProcess[], registryServer: Server) {
    this.procs = procs;
    this.registryServer = registryServer;
  }

  /** Terminate all spawned workers and close the registry. */
  async shutdown(): Promise<void> {
    for (const proc of this.procs) {
      // Negative PID = signal the whole process group (detached: true at
      // spawn time made each child's PID double as its group ID). Falls
      // back to signalling just the direct child if the group signal fails.
      try {
        if (proc.pid) {
          process.kill(-proc.pid, 'SIGTERM');
        } else {
          proc.kill('SIGTERM');
        }
      } catch {
        try { proc.kill('SIGTERM'); } catch { /* ignore */ }
      }
    }
    await this.registryServer.close();
    try { fs.unlinkSync(path.join('.callwire', 'pids')); } catch { /* ignore */ }
  }
}

// ── public API ─────────────────────────────────────────────────────────────

/**
 * Initialise Callwire orchestration.
 *
 * Call once in your application's bootstrap function (e.g., Express startup,
 * NestJS bootstrap, or a standalone `callwire run` CLI).
 *
 * - If `CALLWIRE_SPAWNED=1` → worker mode: register local exports with parent.
 * - Otherwise → orchestrator mode: spawn workers from `callwire.toml`.
 *
 * Returns an `OrchestratorHandle` (orchestrator) or `null` (worker).
 *
 * @example
 * ```ts
 * const handle = await initCallwire();
 * // ... your server logic ...
 * await handle?.shutdown();
 * ```
 */
export async function initCallwire(server?: Server): Promise<OrchestratorHandle | null> {
  if (process.env.CALLWIRE_SPAWNED === '1') {
    await initAsWorker(server);
    return null;
  }
  return initAsOrchestrator();
}

// ── worker mode ────────────────────────────────────────────────────────────

async function initAsWorker(workerServer?: Server): Promise<void> {
  const registryAddr = process.env.CALLWIRE_REGISTRY;
  if (!registryAddr) {
    console.error('[callwire] Worker mode: CALLWIRE_REGISTRY not set — skipping registration');
    return;
  }

  if (!workerServer) {
    throw new Error('initCallwire() in worker mode requires a Server instance');
  }

  // Start the local RPC server on a random port
  const port = await freePort();

  // Re-export everything that was registered via server.export() before init()
  // The Server instance used by the worker IS the global one — the user sets
  // their handlers on it, then calls initCallwire(). We start serving here.
  await workerServer.serve('127.0.0.1', port);
  const workerAddr = `127.0.0.1:${port}`;
  console.error(`[callwire] Worker serving on ${workerAddr}`);

  // Brief grace period
  await new Promise(r => setTimeout(r, 150));

  // Register with parent registry
  const [regHost, regPortStr] = registryAddr.split(':');
  const regClient = new Client();
  await regClient.connect(regHost, Number(regPortStr));

  // The worker exports are registered on _workerHandlers (passed via closure)
  // For TS, the user passes their Server instance — we expose the handler names
  const handlerNames = (workerServer as unknown as { handlers: Map<string, unknown> }).handlers 
    ? Array.from((workerServer as unknown as { handlers: Map<string, unknown> }).handlers.keys())
    : [];

  for (const name of handlerNames) {
    try {
      await regClient.call('callwire.register', [name, workerAddr]);
    } catch (e) {
      console.error(`[callwire] Warning: could not register '${name}': ${e}`);
    }
  }
  regClient.close();
  console.error(`[callwire] Worker registered [${handlerNames.join(', ')}] → ${workerAddr}`);

  // Orphan detection: exit if parent dies
  setInterval(() => {
    try {
      process.kill(process.ppid, 0); // 0 = check existence, throws if dead
    } catch {
      console.error('[callwire] Parent process gone — worker exiting');
      process.exit(0);
    }
  }, 2000);
}

// ── orchestrator mode ──────────────────────────────────────────────────────

async function initAsOrchestrator(): Promise<OrchestratorHandle> {
  const config = parseToml();
  const services = Object.entries(config.services);

  if (services.length === 0) {
    const emptyServer = new Server();
    return new OrchestratorHandle([], emptyServer);
  }

  killStalePids();

  // Start the dynamic registry on a random port
  const regPort = await freePort();
  const registryAddr = `127.0.0.1:${regPort}`;

  // Build the registry using an in-process Server
  const registryStore: Record<string, string[]> = {};
  const registryServer = new Server();

  registryServer.export('callwire.register', ([svcName, addr]) => {
    const key = svcName as string;
    if (!registryStore[key]) registryStore[key] = [];
    if (!registryStore[key].includes(addr as string)) {
      registryStore[key].push(addr as string);
    }
    return null;
  });

  registryServer.export('callwire.discover', ([svcName]) => {
    const addrs = registryStore[svcName as string];
    if (!addrs || addrs.length === 0) {
      throw new Error(`service not found: ${svcName}`);
    }
    return addrs;
  });

  await registryServer.serve('127.0.0.1', regPort);
  console.log(`[callwire] Registry listening on ${registryAddr}`);

  const isProd = (process.env.CALLWIRE_ENV ?? 'dev').toLowerCase() === 'prod';
  const spawnedProcs: child_process.ChildProcess[] = [];

  for (const [name, svc] of services) {
    const cmd = isProd
      ? (svc.prod_cmd || svc.dev_cmd)
      : (svc.dev_cmd || svc.prod_cmd);

    if (!cmd) {
      console.warn(`[callwire] Warning: service '${name}' has no command — skipping`);
      continue;
    }

    const proc = child_process.spawn('sh', ['-c', cmd], {
      env: { ...process.env, CALLWIRE_SPAWNED: '1', CALLWIRE_REGISTRY: registryAddr },
      stdio: 'inherit',
      // Puts the shell AND everything it execs (e.g. "cd x && ./worker")
      // into its own process group (Unix). Without this, shutdown()'s
      // proc.kill() only reaches the "sh -c" wrapper — the actual worker
      // binary underneath is a separate process that never receives the
      // signal, keeps running, and (having inherited stdout/stderr via
      // stdio: 'inherit') keeps those pipes open, hanging anything reading
      // this process's output after it exits. Same root cause as the
      // Python/Go/Rust orchestrators' shell wrapper bug, fixed the same way
      // — shutdown() below signals -proc.pid (the group) instead of proc.pid.
      detached: true,
    });

    proc.on('error', (e) => console.error(`[callwire] Failed to spawn '${name}': ${e.message}`));
    console.log(`[callwire] Spawned '${name}' (PID ${proc.pid}): ${cmd}`);
    spawnedProcs.push(proc);
  }

  writePidFile(spawnedProcs);

  // Wait for workers to come up
  const waitMs = Math.min(1500 * Math.max(spawnedProcs.length, 1), 5000);
  await new Promise(r => setTimeout(r, waitMs));

  console.log(`[callwire] Orchestrator ready — registry at ${registryAddr}`);

  // Graceful shutdown on SIGINT/SIGTERM
  const handle = new OrchestratorHandle(spawnedProcs, registryServer);
  const graceful = () => handle.shutdown().then(() => process.exit(0));
  process.once('SIGINT', graceful);
  process.once('SIGTERM', graceful);

  return handle;
}
